api/sda/sda.go

```diff
@@ -0,0 +1,152 @@
+package sda
+
+import (
+	"bytes"
+	"io"
+	"os"
+	"strconv"
+
+	"github.com/elixir-oslo/crypt4gh/model/headers"
+	"github.com/elixir-oslo/crypt4gh/streaming"
+	"github.com/gofiber/fiber/v2"
+	"github.com/neicnordic/sda-download/internal/config"
+	"github.com/neicnordic/sda-download/internal/database"
+	"github.com/neicnordic/sda-download/pkg/auth"
+	log "github.com/sirupsen/logrus"
+)
+
+// Datasets serves a list of permitted datasets
+func Datasets(c *fiber.Ctx) error {
+
+	// Get access token
+	token, errorCode := auth.GetToken(c.Get("Authorization"))
+	if errorCode > 0 {
+		return fiber.NewError(errorCode, token)
+	}
+
+	// Get permissions
+	visas := c.Locals("visas")
+	datasets := auth.GetPermissions(visas.([]byte))
+	if len(datasets) == 0 {
+		return fiber.NewError(404, "no datasets found")
+	}
+
+	return c.JSON(datasets)
+}
+
+// find looks for a dataset name in a list of datasets
+func find(datasetID string, datasets []string) bool {
+	found := false
+	for i := range datasets {
+		if datasetID == datasets[i] {
+			found = true
+		}
+	}
+	return found
+}
+
+// Files serves file metadata
+func Files(c *fiber.Ctx, datasetID string) error {
+
+	// Get access token
+	token, errorCode := auth.GetToken(c.Get("Authorization"))
+	if errorCode > 0 {
+		return fiber.NewError(errorCode, token)
+	}
+
+	// Get permissions
+	visas := c.Locals("visas")
+	datasets := auth.GetPermissions(visas.([]byte))
+	if len(datasets) == 0 {
+		return fiber.NewError(404, "no datasets found")
+	}
+
+	if find(datasetID, datasets) {
+		// Get file metadata
+		files, err := database.DB.GetFiles(datasetID)
+		if err != nil {
+			// something went wrong with querying or parsing rows
+			log.Errorf("database query failed, %s", err)
+			return c.SendStatus(fiber.StatusInternalServerError)
+		}
+		return c.JSON(files)
+	}
+
+	// no matches in database with given datasetID
+	return c.SendStatus(fiber.StatusUnauthorized)
+}
+
+// Download serves file contents as bytes
+func Download(c *fiber.Ctx, fileID string) error {
+
+	// Get access token
+	token, errorCode := auth.GetToken(c.Get("Authorization"))
+	if errorCode > 0 {
+		return fiber.NewError(errorCode, token)
+	}
+
+	// Get permissions
+	visas := c.Locals("visas")
+	datasets := auth.GetPermissions(visas.([]byte))
+	if len(datasets) == 0 {
+		return fiber.NewError(404, "no datasets found")
+	}
+
+	// Check user has permissions for this file (as part of a dataset)
+	dataset, err := database.DB.CheckFilePermission(fileID)
+	if err != nil {
+		return fiber.NewError(401, "no datasets found with that file ID")
```

Comment:
After propagating this error there is no trace in the logs spitted out by logrus that this actually failed. Why is fiber handling the error here?

Reply:
That error means that the user requested a `fileID`, but there are no `dataset`s that contain such a file.

Added more error logs to HTTP error returns with https://github.com/neicnordic/sda-download/pull/15/commits/c9cf5e4af89a06739599975256e6b5a4bb0eb7f1